### PR TITLE
Add RealmServer method to create a realm dynamically

### PR DESCRIPTION
**Pull request by habdelra**
This PR adds a new method to the `RealmServer` that will dynamically create a realm. As part of this we have also added a new permission: `realm-owner` that owners of realms are granted. Realms are created by default in the `packages/realm-server/realms` directory in the local env.

TODO: 
- [x] username prefix checks
- [x] restart dynamically created realms
- [x] realm-owner permissions tests
- [x] unhappy path tests for dynamically created realms
- [x] Deploy infra PR to add the `MATRIX_REGISTRATION_SHARED_SECRET` env var (https://github.com/cardstack/infra/pull/539)

### CODE REVIEW

File: packages/matrix/tests/registration-with-token.spec.ts [hunk 1]
```diff
@@ -96,6 +106,18 @@ test.describe('User Registration w/ Token', () => {
...
+    page: _page,
+  }) => {
+    // TODO
+  });
+
+  test.skip('it shows an error when the username starts with "_" prefix', async ({
+    page: _page,
+  }) => {
+    // TODO
+  });
```
Comment (habdelra) [inline packages/matrix/tests/registration-with-token.spec.ts]: don't forget this

File: packages/realm-server/main.ts [hunk 2]
```diff
@@ -162,6 +182,7 @@ let dist: URL = new URL(distURL);
 
   await startWorker();
 
+  // TODO also include realms from realms json files on the filesystem
```
Comment (habdelra) [inline packages/realm-server/main.ts]: don't forget this

File: packages/realm-server/tests/permissions/permission-checker-test.ts [hunk 3]
```diff
@@ -8,6 +8,10 @@ let mockMatrixClient = {
     return matrixUserProfile;
   },
 } as MatrixClient;
+
+// TODO add realm-owner tests. the permission currently does not enable any
+// special behavior yet, but we just want to make sure it can be captured
+
```
Comment (habdelra) [inline packages/realm-server/tests/permissions/permission-checker-test.ts]: don't forget

File: packages/realm-server/tests/realm-server-test.ts [hunk 4]
```diff
@@ -2131,24 +2116,200 @@ module('Realm Server', function (hooks) {
...
+        .set('Accept', 'application/vnd.card+json')
+        .set('Authorization', `Bearer ${createJWT(realm, 'rando')}`);
+
+      assert.strictEqual(response.status, 403, 'HTTP 403 status');
+    });
+
+    // TODO can restart a realm that was created dynamically
+    // TODO test that for a server where a realm is mounted at the root, we throw when trying to create a new realm on that server
+    // TODO test that cannot create new realm that collides with existing realm
+    // TODO test that cannot create realm with invalid chars
```
Comment (habdelra) [inline packages/realm-server/tests/realm-server-test.ts]: don't forget

File: packages/realm-server/main.ts (hunk 2 above)
Comment (habdelra) [reply packages/realm-server/main.ts]: done

File: packages/realm-server/tests/realm-server-test.ts (hunk 4 above)
Comment (habdelra) [reply packages/realm-server/tests/realm-server-test.ts]: done

File: packages/realm-server/tests/permissions/permission-checker-test.ts (hunk 3 above)
Comment (habdelra) [reply packages/realm-server/tests/permissions/permission-checker-test.ts]: done

**Review by FadhlanR — APPROVED** [review]
Looks great!

File: packages/matrix/tests/registration-with-token.spec.ts [hunk 5]
```diff
@@ -18,6 +18,16 @@ import { registerUser, createRegistrationToken } from '../docker/synapse';
...
 
+// TODO after we start creating realms as part of user creation we'll need to
+// figure out how to perform better isolation for these tests. Creating a user
+// will result in DB and filesystem state changes that will bleed into the other
+// matrix tests. One idea is that since playwright is running on the server we
+// could perform realm server isolation by using randomly generated DB names's
+// like we do in the realm server tests along with resetting the underlying
+// filesystem and restarting the realm server in between tests that we expect to
+// mutate realm state. These would be rather expensive tests so we should be
+// prudent around the realm server isolation.
```
Comment (lukemelia) [inline packages/matrix/tests/registration-with-token.spec.ts:29]: Maybe we could introduce a virtual (in-memory) filesystem for tests

**Review by lukemelia — APPROVED** [review]

File: packages/matrix/tests/registration-with-token.spec.ts (hunk 5 above)
Comment (habdelra) [reply packages/matrix/tests/registration-with-token.spec.ts:29]: There will still be DB artifacts though that need to be isolated regardless

**Review by backspace — APPROVED** [review]

**Pull request merged**